sdk/dynamicconfig/sdk/metric/sdk.go

```diff
@@ -0,0 +1,625 @@
+) int {
+	checkpointed := 0
+
+	m.current.Range(func(key interface{}, value interface{}) bool {
+		// Note: always continue to iterate over the entire
+		// map by returning `true` in this function.
+		inuse := value.(*record)
+
+		name := inuse.inst.descriptor.Name()
+		shouldCollect := m.shouldCollect(name, periods)
+		if !shouldCollect {
+			return true
+		}
+
+		mods := atomic.LoadInt64(&inuse.updateCount)
+		coll := inuse.collectedCount
+
+		if mods != coll {
+			// Updates happened in this interval,
+			// checkpoint and continue.
+			checkpointed += m.checkpointRecord(inuse)
+			inuse.collectedCount = mods
+			return true
+		}
+
+		// Having no updates since last collection, try to unmap:
+		if unmapped := inuse.refMapped.tryUnmap(); !unmapped {
+			// The record is referenced by a binding, continue.
+			return true
+		}
+
+		// If any other goroutines are now trying to re-insert this
+		// entry in the map, they are busy calling Gosched() awaiting
+		// this deletion:
+		m.current.Delete(inuse.mapkey())
+
+		// There's a potential race between `LoadInt64` and
+		// `tryUnmap` in this function.  Since this is the
+		// last we'll see of this record, checkpoint
+		mods = atomic.LoadInt64(&inuse.updateCount)
+		if mods != coll {
+			checkpointed += m.checkpointRecord(inuse)
+		}
+		return true
+	})
+
+	return checkpointed
+}
+
+// CollectAsync implements internal.AsyncCollector.
+func (m *Accumulator) CollectAsync(kv []kv.KeyValue, obs ...metric.Observation) {
+	labels := label.NewSetWithSortable(kv, &m.asyncSortSlice)
+
+	for _, ob := range obs {
+		if a := m.fromAsync(ob.AsyncImpl()); a != nil {
+			a.observe(ob.Number(), &labels)
+		}
+	}
+}
+
+func (m *Accumulator) observeAsyncInstruments(
+	ctx context.Context,
+	periods []pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod,
+) int {
+	m.asyncLock.Lock()
+	defer m.asyncLock.Unlock()
+
+	asyncCollected := 0
+
+	// TODO: change this to `ctx` (in a separate PR, with tests)
+	m.asyncInstruments.Run(context.Background(), m)
+
+	for _, inst := range m.asyncInstruments.Instruments() {
+		if a := m.fromAsync(inst); a != nil &&
+		m.shouldCollect(a.descriptor.Name(), periods) {
+			asyncCollected += m.checkpointAsync(a)
+		}
+	}
+
+	return asyncCollected
+}
+
+// Determines if we should collect metrics from an instrument with `name`.
+func (m *Accumulator) shouldCollect(
+	name string,
+	periods []pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod,
+) bool {
+	// If we aren't enabling the dynamic config extension, we collect everything.
+	if m.ext == nil {
+		return true
+	}
+
+	m.ext.lock.Lock()
+	// Determine CollectionPeriod associated with instrument.
+	var collectionPeriod pb.ConfigResponse_MetricConfig_Schedule_CollectionPeriod = 0
+	if period, ok := m.ext.instrumentPeriod[name]; ok {
+		collectionPeriod = period
+	} else {
+		collectionPeriod = FindPeriod(name, m.ext.schedules)
+		m.ext.instrumentPeriod[name] = collectionPeriod
+	}
+	m.ext.lock.Unlock()
```

Comment:
Since the mutex belongs to the DynamicExtension struct, would it be possible to move this logic into the DynamicExtension struct, and use its lock within the method? Otherwise, if external code starts messing with the mutex, the threading model may be more difficult to reason about, as Ming mentioned earlier?

Reply:
Sorry about this, I should've pushed revisions earlier.